torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -758,147 +773,25 @@ def sort_tokens(self, x, topk_ids, topk_weights):
-            m_actual_tokens = valid_tokens.shape[0]
-            intermediate_size = self.get_parameter("gate_proj_weight").shape[1]
-            hidden_size = self.get_parameter("down_proj_weight").shape[1]
-
-            # Prepare input in FP8 format (shared by gate and up projections)
-            if self.use_triton_quant:
-                gate_up_input = dsgemm_kernels.groupwise_activation_quant(
-                    valid_tokens
-                )  # dsgemm_kernels.activation_quant_triton(valid_tokens)
-            else:
-                gate_up_input = dsgemm_utils.prepare_fp8_input(valid_tokens)
 
-            # Allocate output buffers
-            gate_proj_out = torch.empty(
-                (m_actual_tokens, intermediate_size),
-                device=contig_tokens.device,
-                dtype=contig_tokens.dtype,
-            )
-            up_proj_out = torch.empty_like(gate_proj_out)
-
-            # Run first GEMM (gate projection)
-            deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_contiguous(
-                gate_up_input,
-                (gate_proj_weight_fp8, gate_proj_scales),
-                gate_proj_out,
-                m_indices,
-            )
+    def _run_group_gemm(self, contig_tokens, m_sizes, m_offsets):
+        """Run the appropriate group GEMM implementation based on configuration"""
 
-            # Run second GEMM (up projection)
-            deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_contiguous(
-                gate_up_input,
-                (up_proj_weight_fp8, up_proj_scales),
-                up_proj_out,
-                m_indices,
+        try:
+            return self.group_gemm_strategies[self.group_mm].execute(
+                contig_tokens, m_sizes, m_offsets, self
             )
-
-            # Apply activation
-            hidden_states = MLP.act_fn(gate_proj_out) * up_proj_out
-
-            # Allocate output buffer for down projection
-            down_proj_out = torch.empty(
-                (m_actual_tokens, hidden_size),
-                device=contig_tokens.device,
-                dtype=contig_tokens.dtype,
-            )
-
-            # Run third GEMM (down projection)
-            if self.use_triton_quant:
-                hidden_states_quantized = dsgemm_kernels.groupwise_activation_quant(
-                    hidden_states
-                )
-            else:
-                hidden_states_quantized = dsgemm_utils.prepare_fp8_input(hidden_states)
-
-            deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_contiguous(
-                hidden_states_quantized,
-                (down_proj_weight_fp8, down_proj_scales),
-                down_proj_out,
-                m_indices,
-            )
-
-            # Copy results back to contig_tokens
-            contig_tokens[: m_offsets[-1]] = down_proj_out
-            return contig_tokens
+        except Exception as e:
+            # Flag the error
+            print(f"Error using {self.group_mm} strategy: {e}")
```

Comment:
No need for the try-except it seem.

Reply:
actually it has been very helpful isolating issues working with the various group gemms.  It provides much better error clarity if things go wrong.